ts/receiver/receiver.ts

```diff
@@ -0,0 +1,477 @@
+        return false;
+    }
+    if (!authorisation.grantSignature) {
+        return false;
+    }
+    const isValid = await window.libloki.crypto.validateAuthorisation(
+        authorisation
+    );
+    if (!isValid) {
+        return false;
+    }
+    const correctSender = pubKey === authorisation.secondaryDevicePubKey;
+    if (!correctSender) {
+        return false;
+    }
+    const { primaryDevicePubKey } = authorisation;
+    // ensure the primary device is a friend
+    const c = window.ConversationController.get(primaryDevicePubKey);
+    if (!c || !await c.isFriendWithAnyDevice()) {
+        return false;
+    }
+    await window.libloki.storage.savePairingAuthorisation(authorisation);
+
+    return true;
+}
+
+async function handleAutoFriendRequestMessage(
+    source: string,
+    ourPubKey: string,
+    conversation: ConversationModel
+) {
+    const isMe = source === ourPubKey;
+    // If we got a friend request message (session request excluded) or
+    // if we're not friends with the current user that sent this private message
+    // Check to see if we need to auto accept their friend request
+    if (isMe) {
+        window.log.info('refusing to add a friend request to ourselves');
+        throw new Error('Cannot add a friend request for ourselves!');
+    }
+    // auto-accept friend request if the device is paired to one of our friend's primary device
+    const shouldAutoAcceptFR = await handleSecondaryDeviceFriendRequest(source);
+    if (shouldAutoAcceptFR) {
+        window.libloki.api.debug.logAutoFriendRequest(
+            `Received AUTO_FRIEND_REQUEST from source: ${source}`
+        );
+        // Directly setting friend request status to skip the pending state
+        await conversation.setFriendRequestStatus(
+            window.friends.friendRequestStatusEnum.friends
+        );
+        // sending a message back = accepting friend request
+
+        window.libloki.api.sendBackgroundMessage(
+            source,
+            window.textsecure.OutgoingMessage.DebugMessageType.AUTO_FR_ACCEPT
+        );
+        // return true to notify the message is fully processed
+        return true;
+    }
+    return false;
+}
+
+export async function handle_message_event(event: any): Promise<void> {
+    const { data, confirm } = event;
+
+    const isIncoming = event.type === 'message';
+
+    if (!data || !data.message) {
+        window.log.warn('Invalid data passed to handle_message_event.', event);
+        confirm();
+        return;
+    }
+
+    const { message, destination } = data;
+
+    let { source } = data;
+
+    const isGroupMessage = message.group;
+
+    const type = isGroupMessage ? ConversationType.GROUP : ConversationType.PRIVATE;
+
+    // MAXIM: So id is actually conversationId
+    const id = isIncoming ? source : destination;
+
+    const {
+        PROFILE_KEY_UPDATE,
+        SESSION_REQUEST,
+        SESSION_RESTORE,
+    } = window.textsecure.protobuf.DataMessage.Flags;
```

Comment:
`window.textsecure.protobuf` can be replaced with `SignalService` directly

Reply:
done